syntax = "proto3";

package dronecorerpc;

option java_package = "io.dronecore.rpc";
option java_outer_classname = "MissionProto";

service MissionRPC {
    rpc SendMission(SendMissionRequest) returns(MissionResult) {}
    rpc StartMission(StartMissionRequest) returns(MissionResult) {}
}

message UUID {
    uint64 uuid = 1;
}

message SendMissionRequest {
    UUID uuid = 1;
    Mission mission = 2;
}

message Mission {
    repeated MissionItem mission_items = 1;
}

message MissionItem {
    double latitude_deg = 1;
    double longitude_deg = 2;
    double relative_altitude_m = 3;
    float speed_m_s = 4;
    bool is_fly_through = 5;
    float gimbal_pitch_deg = 6;
    float gimbal_yaw_deg = 7;
    CameraAction camera_action = 8;

    enum CameraAction {
        NONE = 0;
        TAKE_PHOTO = 1;
        START_PHOTO_INTERVAL = 2;
        STOP_PHOTO_INTERVAL = 3;
        START_VIDEO = 4;
        STOP_VIDEO = 5;
    }
}

message MissionResult {
    Result result = 1;
    string result_str = 2;

    enum Result {
        SUCCESS = 0;
        ERROR = 1;
        TOO_MANY_MISSION_ITEMS = 2;
        BUSY = 3;
        TIMEOUT = 4;
        INVALID_ARGUMENT = 5;
        UNKNOWN = 6;
    }
}

message StartMissionRequest {
    UUID uuid = 1;
}
